### ISSUE TITLE
Icinga 2 stalls and no further updates are populate to IDO, IcingaDB and InfluxDB

### ISSUE THREAD

**Issue by mcktr** [Icinga/icinga2#8160]
## Describe the bug

I run Icinga 2 v2.12.0-1 in my test environment on a virtual machine. I also installed Icinga Web 2, IcingaDB, Redis, InfluxDB and Grafana on this VM. After some time the Icinga 2 process stalls and no further updates are show in Icinga Web 2 and Grafana. The Icinga Web 2 shows the message that the backend is not running. The log and debuglog does not indicate anything it just stopps.

## To Reproduce

I use a very simple setup with no agents/satellites. The VM is configured as main node using the `icinga2 node wizard` and selecting the master setup.

I simulate host and service objects with the following configuration:

```
# vim /etc/icinga2/zones.d/master/many.conf 

const countHosts = 250
const countServices = 10

for (i in range(countHosts)) {
	object Host "many-test-" + i use(i) {
		check_command = "random"
		address = "127.0.0.1"		
	}
}

for (j in range(countServices)) {
	apply Service "many-test-" + j {
		check_command = "random"

		check_interval = 30s
		retry_interval = 30s

		assign where match("many*", host.name)	
	}
}
```

I also added self monitoring for Icinga 2:

```
# vim /etc/icinga2/zones.d/master/self.conf 

object Host "deb10-icinga" {
	address = "127.0.0.1"

	check_command = "dummy"
}

apply Service "load" {
	check_command = "load"

	check_interval = 10s
	retry_interval = 10s

	assign where match("deb10-icinga", host.name)
}

apply Service "procs" {
	check_command = "procs"

	check_interval = 10s
	retry_interval = 10s

	assign where match("deb10-icinga", host.name)
}

apply Service "swap" {
	check_command = "swap"

	check_interval = 10s
	retry_interval = 10s

	assign where match("deb10-icinga", host.name)
}

apply Service "disk" {
	check_command = "disk"

	check_interval = 10s
	retry_interval = 10s

	assign where match("deb10-icinga", host.name)
}

apply Service "icinga" {
	check_command = "icinga"

	check_interval = 10s
	retry_interval = 10s

	assign where match("deb10-icinga", host.name)
}

apply Service "ido" {
	check_command = "ido"

	check_interval = 10s
	retry_interval = 10s

	vars.ido_type = "IdoMysqlConnection"
	vars.ido_name = "ido-mysql"

	assign where match("deb10-icinga", host.name)
}

```

After waiting some time (30 minutes to ~2 hours) Icinga 2 stalls and no new updates/check results are populated too Icinga Web 2 and InfluxDB/Grafana.

## Expected behavior

The Icinga 2 process keeps running and does not stall at some point.

## Screenshots

![Bildschirmfoto von 2020-08-09 18-39-40](https://user-images.githubusercontent.com/18580278/89737567-13b55800-da72-11ea-8114-01aa5f5ae8d7.png)
![Bildschirmfoto von 2020-08-09 18-39-46](https://user-images.githubusercontent.com/18580278/89737569-157f1b80-da72-11ea-9055-73e595fffe1f.png)
![Bildschirmfoto von 2020-08-09 18-39-50](https://user-images.githubusercontent.com/18580278/89737570-16b04880-da72-11ea-9954-291f49cc53f3.png)

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`): r2.12.0-1
* Operating System and version: Debian GNU/Linux 10 (buster)
* Enabled features (`icinga2 feature list`): api checker debuglog icingadb ido-mysql influxdb mainlog notification
* Icinga Web 2 version and modules (System - About): 2.8.1 (icingadb: 1.0.0-rc1, ipl: v0.5.0, monitoring: 2.8.1)
* Config validation (`icinga2 daemon -C`): 
```
[2020-08-09 18:43:14 +0200] information/cli: Icinga application loader (version: r2.12.0-1)
[2020-08-09 18:43:14 +0200] information/cli: Loading configuration file(s).
[2020-08-09 18:43:14 +0200] information/ConfigItem: Committing config item(s).
[2020-08-09 18:43:14 +0200] information/ApiListener: My API identity: deb10-icinga
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 NotificationComponent.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 251 Hosts.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 2 FileLoggers.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 IcingaApplication.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 IcingaDB.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 CheckerComponent.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 3 Zones.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 Endpoint.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 IdoMysqlConnection.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 2 ApiUsers.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 ApiListener.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 235 CheckCommands.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 1 InfluxdbWriter.
[2020-08-09 18:43:14 +0200] information/ConfigItem: Instantiated 2506 Services.
[2020-08-09 18:43:14 +0200] information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
[2020-08-09 18:43:14 +0200] information/cli: Finished validating the configuration file(s).

```
* ~~If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.~~

## Additional context

I did a quick dive into debugging and I think at some point all threads are in a waiting state and/or are waiting for locks to be freed. I have another environment with the same versions but with the difference that there is no IcingaDB installed/activated, so just a guess from my side somewhere in the IcingaDB implementation occurs a dead lock.

Here you'll find a full GDB backtrace:

[gdb_bt_5949_1596989893.log](https://github.com/Icinga/icinga2/files/5048092/gdb_bt_5949_1596989893.log)
[gdb_bt_5969_1596989876.log](https://github.com/Icinga/icinga2/files/5048093/gdb_bt_5969_1596989876.log)
[gdb_bt_5994_1596989864.log](https://github.com/Icinga/icinga2/files/5048094/gdb_bt_5994_1596989864.log)

**Comment by N-o-X** [comment]
Hi @mcktr,

thanks for testing! I'm able to reproduce this issue on MacOS. This problem seems to be Service::GetSeverity(). Here we first lock the service and then its host. It's possible that we do it the other way around (Host first, then service) somewhere else in the code, which then creates a deadlock (both ways at the same time). While I'm still looking for this mentioned counterpart, #7905 could help us in the long run.

Greetings
Noah

**Comment by Al2Klimov** [comment]
And on my CentOS Icinga 2 freezes completely:

```
[2020-08-10 08:18:42 -0700] notice/ApiListener: Relaying 'event::SetNextCheck' message
[2020-08-10 08:18:42 -0700
```

**Comment by Al2Klimov** [comment]
_sigh_

```
(gdb) thread apply all bt

Thread 20 (Thread 0x7fcc7cfbd700 (LWP 20979)):
#0  0x00007fcc7a2ec54d in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fcc7a2e7eb6 in _L_lock_941 () from /lib64/libpthread.so.0
#2  0x00007fcc7a2e7daf in pthread_mutex_lock () from /lib64/libpthread.so.0
#3  0x00000000007eb236 in __gthread_mutex_lock (__mutex=<optimized out>) at /usr/include/c++/4.8.2/x86_64-redhat-linux/bits/gthr-default.h:748
#4  __gthread_recursive_mutex_lock (__mutex=<optimized out>) at /usr/include/c++/4.8.2/x86_64-redhat-linux/bits/gthr-default.h:810
#5  lock (this=<optimized out>) at /usr/include/c++/4.8.2/mutex:175
#6  Lock (this=0x7fcc7cfbba80) at ../base/objectlock.cpp:32
#7  __base_ctor  (this=0x7fcc7cfbba80, object=<optimized out>) at ../base/objectlock.cpp:25
#8  icinga::ObjectLock::ObjectLock(boost::intrusive_ptr<icinga::Object> const&) (this=this@entry=0x7fcc7cfbba80, object=...) at ../base/objectlock.cpp:17
#9  0x0000000000ac7413 in icinga::Service::GetSeverity (this=0x7fcc68195020) at ../icinga/service.cpp:139
```

**Comment by Al2Klimov** [comment]
Also affects RC1.

**Comment by yoshi314** [comment]
are check processes stuck in defunct state? i am observing something like this on my secondary master node in HA setup. it happens when i reload config on main node and its propagated to secondary master.

**Comment by Al2Klimov** [comment]
Even b6e0abc64 is affected.

**Comment by Al2Klimov** [comment]
e55c15d doesn't help.

**Comment by yoshi314** [comment]
so far my workaround is as in linked issue. I have two nodes in one zone.

node that receives configuration is on 2.12, the node that cfg replicates to is on 2.11. Even though the main node receives also cfg from director, it's not affected somehow.

Main node is on debian8, secondary node is on debian10. Things are running fine at the moment.

**Comment by N-o-X** [comment]
Closed by #8168
